lib/decorateWithMetrics.js

```diff
@@ -0,0 +1,177 @@
+                methodName,
+                total,
+                max,
+                min,
+                avg,
+                calls,
+            }];
+        })
+        .object() // Create a map like methodName -> StatSummary
+        .value();
+
+    const totalTime = sum(_.values(summaries), 'total');
+    const averageTime = (totalTime / _.size(summaries)) || 0;
+
+    return {
+        totalTime,
+        averageTime,
+        summaries,
+    };
+}
+
+function toHumanReadableDuration(millis) {
+    const minute = 60 * 1000;
+    if (millis > minute) {
+        return `${(millis / minute).toFixed(1)}min`;
+    }
+
+    const second = 1000;
+    if (millis > second) {
+        return `${(millis / second).toFixed(2)}sec`;
+    }
+
+    return `${millis.toFixed(3)}ms`;
+}
+
+/**
+ * Print extensive information on the dev console
+ * max, min, average, total time for each method
+ * and a table of individual calls
+ */
+function printMetrics() {
+    const {totalTime, averageTime, summaries} = getMetrics();
+
+    /* eslint-disable no-console */
+    console.group('Onyx Benchmark');
+    console.info('  Total: ', toHumanReadableDuration(totalTime));
+    console.info('  Average: ', toHumanReadableDuration(averageTime));
+
+    _.chain(summaries)
+        .sortBy('avg')
+        .reverse()
+        .forEach(({calls, methodName, ...summary}) => {
+            const times = _.map(summary, (value, key) => `${key}: ${toHumanReadableDuration(value)}`);
+
+            console.groupCollapsed(`${methodName}\n  ${times.join('\n  ')} \n  calls: ${calls.length}`);
+            console.table(calls.map(call => ({
+                startTime: toHumanReadableDuration(call.startTime),
+                endTime: toHumanReadableDuration(call.endTime),
+                duration: toHumanReadableDuration(call.duration),
+                args: JSON.stringify(call.args)
+            })));
+            console.groupEnd();
+        });
+    console.groupEnd();
+    /* eslint-enable */
+}
+
+/**
+ * Clears all collected metrics.
+ */
+function resetMetrics() {
+    stats = {};
+}
+
+/**
+ * @typedef {Object} CallMeta
+ * @property {string} methodName
+ * @property {number} startTime - start time in ms relative to app startup - 0.00 is exactly at start
+ * @property {number} endTime - end time in ms relative to app startup - 0.00 is exactly at start
+ * @property {number} [duration] - the it took to resolve this async call
+ * @property {Array<any>} args - arguments passed to the call
+ */
+
+/**
+ * @typedef {Object} StatSummary
+ * @property {string} methodName
+ * @property {number} minTime - the quickest execution for this method
+ * @property {number} maxTime - the longest execution for this method
+ * @property {number} averageTime - average execution time for the method
+ * @property {number} totalTime - sum of the duration for all the calls
+ * @property {Array<CallMeta>} calls - a collection of all the calls for this method
+ */
```

Comment:
What are all these docs down here for? I think I would just remove them.

Reply:
I needed these while I was writing tests, I don't need them any more, will remove

---

These help with the editor's intellisense:

#### Hovering summaries
![image](https://user-images.githubusercontent.com/12156624/119379438-2844a300-bcc8-11eb-9105-67abb5c143a5.png)

#### Aftering clicking StatSummary
![image](https://user-images.githubusercontent.com/12156624/119379501-38f51900-bcc8-11eb-8ad3-d7dbafcc1fba.png)

#### And after clicking CallMeta
![image](https://user-images.githubusercontent.com/12156624/119379937-bde03280-bcc8-11eb-822c-71a892347945.png)

```js
/**
 * Returns total, average time and all captured stats mapped under
 * summaries.methodName -> StatSummary
 * @returns {{averageTime: number, summaries: Record<string, StatSummary>, totalTime: number}}
 */
```

The same can be achieved by making a separate file defining the public interface, libraries tend to have them
Here's the one for AsyncStorage: https://github.com/react-native-async-storage/async-storage/blob/master/types/index.d.ts

Reply:
Ah, OK. I learned something new!